File: src/main/python/AoC2023_03.py
#! /usr/bin/env python3
#
# Advent of Code 2023 Day 3
#

import re
import sys
from collections import defaultdict
from math import prod
from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import Cell
from aoc.grid import CharGrid

TEST = """\
467..114..
...*......
..35..633.
......#...
617*......
.....+.58.
..592.....
......755.
...$.*....
.664.598..
"""


class EnginePart(NamedTuple):
    part: str
    number: int
    pos: Cell


Input = list[EnginePart]
Output1 = int
Output2 = int


class Solution(SolutionBase[list[EnginePart], Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        grid = CharGrid([[c for c in line] for line in input_data])

        def find_engine_part(
            row: int, col_span: tuple[int, int]
        ) -> EnginePart | None:
            for col in range(*col_span):
                for n in grid.get_all_neighbours(Cell(row, col)):
                    val = grid.get_value(n)
                    if not val.isdigit() and val != ".":
                        s = grid.get_row_as_string(row)[
                            col_span[0] : col_span[1]  # noqa E203
                        ]
                        return EnginePart(val, int(s), n)
            return None

        return [
            ep
            for ep in (
                find_engine_part(r, m.span())
                for r, row in enumerate(grid.get_rows_as_strings())
                for m in re.finditer(r"[0-9]+", row)
            )
            if ep is not None
        ]

    def part_1(self, engine_parts: list[EnginePart]) -> Output1:
        return sum(ep.number for ep in engine_parts)

    def part_2(self, engine_parts: list[EnginePart]) -> Output2:
        d = defaultdict(list)
        for gear in (ep for ep in engine_parts if ep.part == "*"):
            d[gear.pos].append(gear.number)
        return sum(prod(v) for v in d.values() if len(v) == 2)

    @aoc_samples(
        (
            ("part_1", TEST, 4361),
            ("part_2", TEST, 467835),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2023, 3)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2023_03.java
import static com.github.pareronia.aoc.Utils.enumerate;
import static java.util.stream.Collectors.groupingBy;

import java.util.List;
import java.util.Optional;
import java.util.regex.Pattern;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.Range;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2023_03
        extends SolutionBase<List<AoC2023_03.EnginePart>, Integer, Integer> {
    
    private AoC2023_03(final boolean debug) {
        super(debug);
    }
    
    public static AoC2023_03 create() {
        return new AoC2023_03(false);
    }
    
    public static AoC2023_03 createDebug() {
        return new AoC2023_03(true);
    }
    
    @Override
    protected List<EnginePart> parseInput(final List<String> inputs) {
        final Pattern regex = Pattern.compile("[0-9]+");
        final CharGrid grid = CharGrid.from(inputs);
        return enumerate(grid.getRowsAsStrings())
            .flatMap(e -> regex.matcher(e.getValue()).results()
                .map(m -> findEnginePart(grid, e.getIndex(),
                                         Range.between(m.start(), m.end()))))
            .filter(Optional::isPresent)
            .map(Optional::get)
            .toList();
    }
    
    private Optional<EnginePart> findEnginePart(
            final CharGrid grid, final int row, final Range<Integer> colspan
    ) {
        return IntStream.range(colspan.getMinimum(), colspan.getMaximum())
            .boxed()
            .flatMap(col -> grid.getAllNeighbours(Cell.at(row, col)))
            .filter(n -> !Character.isDigit(grid.getValue(n)))
            .filter(n -> grid.getValue(n) != '.')
            .map(n -> {
                final String s = grid.getRowAsString(row)
                            .substring(colspan.getMinimum(), colspan.getMaximum());
                return new EnginePart(grid.getValue(n), Integer.parseInt(s), n);
            })
            .findFirst();
    }

    @Override
    public Integer solvePart1(final List<EnginePart> engineParts) {
        return engineParts.stream().mapToInt(EnginePart::number).sum();
    }
    
    @Override
    public Integer solvePart2(final List<EnginePart> engineParts) {
        return engineParts.stream()
            .filter(ep -> ep.part() == '*')
            .collect(groupingBy(EnginePart::pos))
            .values().stream()
            .filter(v -> v.size() == 2)
            .mapToInt(v -> v.get(0).number() * v.get(1).number())
            .sum();
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "4361"),
        @Sample(method = "part2", input = TEST, expected = "467835"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2023_03.create().run();
    }

    private static final String TEST = """
            467..114..
            ...*......
            ..35..633.
            ......#...
            617*......
            .....+.58.
            ..592.....
            ......755.
            ...$.*....
            .664.598..
            """;
    
    record EnginePart(char part, int number, Cell pos) { }
}
